platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1111,21 +1105,45 @@ public void collectVariants(VariantVisitor visitor) {
 
     @Override
     public boolean isCanBeMutated() {
-        return canBeMutated;
+        boolean immutable = observed || currentResolveState.get().isPresent();
+        return !immutable;
     }
 
     @Override
-    public void preventFromFurtherMutation() {
-        if (!canBeMutated) {
+    public void markAsObserved() {
+        if (observed) {
             return;
         }
 
-        // TODO This should use the same `MutationValidator` infrastructure that we use for other mutation types
+        runActionInHierarchy(conf -> {
+            conf.configurationAttributes.freeze();
+            conf.outgoing.preventFromFurtherMutation();
+            conf.preventUsageMutation();
+            conf.observed = true;
```

Comment:
❓ Shouldn't this be checking `if (!conf.observed)` before acting on the current conf?

Reply:
All actions in the callback are idempotent, so the existing implementation is not incorrect. The initial `if (observed)` is an optimizations that will likely prevent most unnecessary calls of the idempotent methods, but I can add this to be extra sure we're not running these over and over

Reply:
Yeah, even if they don't do anything, still worth not calling them if unnecessary.